include/dlaf/matrix/copy_tile.h

```diff
@@ -12,14 +12,73 @@
 
 #include <type_traits>
 
+#if DLAF_WITH_CUDA
+#include <cuda_runtime.h>
+
+#include "dlaf/cuda/error.h"
+#endif
+
 #include "dlaf/lapack_tile.h"
 #include "dlaf/matrix/tile.h"
 
 namespace dlaf {
+namespace internal {
+template <typename T, Device Source, Device Destination>
+struct CopyTile;
+
+template <typename T>
+struct CopyTile<T, Device::CPU, Device::CPU> {
+  static void call(const matrix::Tile<const T, Device::CPU>& source,
+                   const matrix::Tile<T, Device::CPU>& destination) {
+    dlaf::tile::lacpy<T>(source, destination);
+  }
+};
+
+template <typename T>
+struct CopyTile<T, Device::CPU, Device::GPU> {
+  static void call(const matrix::Tile<const T, Device::CPU>& source,
+                   const matrix::Tile<T, Device::GPU>& destination) {
+    SizeType m = source.size().rows();
+    SizeType n = source.size().cols();
+
+    DLAF_CUDA_CALL(cudaMemcpy2D(destination.ptr(), destination.ld() * sizeof(T), source.ptr(),
+                                source.ld() * sizeof(T), m * sizeof(T), n, cudaMemcpyDefault));
+  }
+};
 
-template <class T>
-void copy(const matrix::Tile<const T, Device::CPU>& source, const matrix::Tile<T, Device::CPU>& dest) {
-  dlaf::tile::lacpy<T>(source, dest);
+template <typename T>
+struct CopyTile<T, Device::GPU, Device::CPU> {
+  static void call(const matrix::Tile<const T, Device::GPU>& source,
+                   const matrix::Tile<T, Device::CPU>& destination) {
+    SizeType m = source.size().rows();
+    SizeType n = source.size().cols();
+
+    DLAF_CUDA_CALL(cudaMemcpy2D(destination.ptr(), destination.ld() * sizeof(T), source.ptr(),
+                                source.ld() * sizeof(T), m * sizeof(T), n, cudaMemcpyDefault));
+  }
+};
+
+template <typename T>
+struct CopyTile<T, Device::GPU, Device::GPU> {
+  static void call(const matrix::Tile<const T, Device::GPU>& source,
+                   const matrix::Tile<T, Device::GPU>& destination) {
+    SizeType m = source.size().rows();
+    SizeType n = source.size().cols();
+
+    DLAF_CUDA_CALL(cudaMemcpy2D(destination.ptr(), destination.ld() * sizeof(T), source.ptr(),
+                                source.ld() * sizeof(T), m * sizeof(T), n, cudaMemcpyDefault));
+  }
+};
 }
 
+template <typename T, Device Source, Device Destination>
+void copy(const matrix::Tile<const T, Source>& source, const matrix::Tile<T, Destination>& destination) {
+  DLAF_ASSERT_HEAVY(
+      source.size().rows() <= destination.size().rows(),
+      "number of rows in destination must be at least as large as number of rows in source for tile copy");
+  DLAF_ASSERT_HEAVY(
+      source.size().cols() <= destination.size().cols(),
+      "number of columns in destination must be at least as large as number of columns in source for tile copy");
```

Comment:
I'm not sure we should allow copy to larger tiles... What is the use case you have in mind?

Reply:
None. I left it like this only because it's possible. I'll happily make it stricter if you prefer that.

Reply:
I've updated the assertions to require equality.